README
NAME
    Text::KyTea - Perl wrapper for KyTea

SYNOPSIS
      use Text::KyTea;

      my $kytea   = Text::KyTea->new(%config);
      my $results = $kytea->parse($text);

      for my $result (@{$results})
      {
          print $result->{surface};

          for my $tags (@{$result->{tags}})
          {
              print "\t";

              for my $tag (@{$tags})
              {
                  print " ", $tag->{feature}, "/", $tag->{score};
              }
          }

          print "\n";
      }

DESCRIPTION
    KyTea is a general toolkit developed for analyzing text, with a focus on
    Japanese, Chinese and other languages requiring word or morpheme
    segmentation.

    This module works under KyTea Ver.0.4.2 and later. Under the old
    versions of KyTea, this module does not work!

    If you've changed the default install directory of KyTea, please install
    Text::KyTea in interactive mode (e.g., cpanm --interactive or cpanm -v).

    For more information about KyTea, please see the "SEE ALSO" section of
    this page.

METHODS
  $kytea = Text::KyTea->new(%config)
    Creates a new Text::KyTea instance.

      my $kytea = Text::KyTea->new(
          model       => 'model.bin', # default is "$INSTALL_DIRECTORY/share/kytea/model.bin"
          notag       => [1,2],       # default is []
          nounk       => 0,           # default is 0 (Estimates the pronunciation of unkown words.)
          unkbeam     => 50,          # default is 50
          tagmax      => 3,           # default is 3
          deftag      => 'UNK',       # default is 'UNK'
          unktag      => '',          # default is ''
          prontag_num => 1,           # default is 1 (Normally no need to change the default value.)
      );

  $results_arrayref = $kytea->parse($text)
    Parses the given text via KyTea, and returns the results. The results
    are returned as an array reference.

  $pron = $kytea->pron( $text [, $replacement ] )
    Returns the estimated pronunciation of the given text. Unknown
    pronunciations are replaced with $replacement.

    If $replacement is not specified, unknown pronunciations are replaced
    with the original characters.

    This is the mere shortcut method.

  $kytea->read_model($path)
    Reads the given model file. The model file should be read by new(model
    => $path) method.

    Model files are available at <http://www.phontron.com/kytea/model.html>.

AUTHOR
    pawa <pawapawa@cpan.org>

SEE ALSO
    <http://www.phontron.com/kytea/>

LICENSE
    Copyright (C) pawa All rights reserved.

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Text::KyTea

0.43    Wed, 09 Jan 2013 21:36:57 +0900
        - Fixed typo

0.42    Wed, 09 Jan 2013 18:29:15 +0900
        - Added pron method
        - Improved Makefile.PL

0.41    Fri, 28 Sep 2012 17:10:46 +0900
        - improved Makefile.PL

0.40    Sun, 03 Jun 2012 20:48:11 +0900
        (This version supports KyTea Ver.0.4.2 and later.)

        - removed h2z option (KyTea Ver.0.4.2 and later support this function)
          - removed dependencies on Data::Recursive and Lingua::JA::Regular::Unicode
        - replaced Test::Extension with Test::Fatal

0.31    Mon, 14 May 2012 19:25:21 +0900
        - fixed documentation
        - improved namespace pollutions

0.30    Fri, 10 Feb 2012 18:23:29 +0900
        - added h2z option (thanks to Shinsuke MORI of Kyoto University)
        - added t/03_h2z.t, t/04_parse_h2z.t, t/08_exception.t
        - modified typemap
        - improved namespace pollution
        - added train_kytea subroutine to Makefile.PL
        - added dependency on Data::Recursive::Encode,
        Lingua::JA::Regular::Unicode, Test::Exception

0.23_1  Fri, 03 Feb 2012 10:28:57 +0900
        - removed write_model_method
        - updated model/test.mod from KyTea Ver.0.3.x model to KyTea Ver.0.4.x model

0.23    Sun, 06 Nov 2011 09:27:14 +0900
        - removed build dependency on Devel::CheckLib

0.22    Sun, 06 Nov 2011 02:31:25 +0900
        - fixed Makefile.PL
        - fixed POD
        - cleaned some redundant code

0.21    Mon, 24 Oct 2011 16:58:57 +0900
        - fixed POD

0.20    Sun, 23 Oct 2011 22:31:54 +0900
        - added options
        - added write_model method

0.13    Wed, 19 Oct 2011 12:32:58 +0900
        - fixed leaktrace.t
        - fixed typemap
        - fixed parse.t
        - changed test order

0.12    Tue, 18 Oct 2011 14:41:55 +0900
        - fixed typemap
        - fixed parse.t

0.11    Mon, 17 Oct 2011 19:20:40 +0900
        - cleaned some redundant code

0.10    Mon, 17 Oct 2011 03:03:27 +0900
        - changed API
        - fixed POD, Makefile.PL, basic.t
        - added leaktrace.t, parse.t, model/test.mod, read_model method
        - moved some files

0.04    Sat, 17 Sep 2011 21:25:41 +0900
        - fixed Makefile.PL

0.03    Fri Sep 16 22:31:13 2011 +0900
        - fixed collisions between C++ macros and functions defined by perl (thanks to perlcodesample)
        - added valgrind.t

0.02    Sun Sep 11 12:42:10 2011 +0900
        - fixed pod (thanks to tokuhirom)

0.01    Tue Sep  6 02:09:33 2011
        - original version


Makefile.PL
use strict;
use warnings;
use inc::Module::Install;
use Module::Install::XSUtil;

my $MIN_VERSION     = '0.4.2';
my $MIN_VERSION_INT = '42';

my $INSTALL_DIR = ExtUtils::MakeMaker::prompt('Input the path to the install directory of KyTea', '/usr/local');
my $LIB_PATH    = "${INSTALL_DIR}/lib";
my $INC_PATH    = "${INSTALL_DIR}/include";
my $BIN_PATH    = "${INSTALL_DIR}/bin";

my $TEST_MODEL_DIR  = './model';
my $TEST_MODEL_FILE = "${TEST_MODEL_DIR}/test.txt";
my $TEST_MODEL      = "${TEST_MODEL_DIR}/test.mod";

all_from    'lib/Text/KyTea.pm';
readme_from 'lib/Text/KyTea.pm';

requires       'Carp';
build_requires 'Devel::CheckLib' => 0.98;
test_requires  'Test::More'      => 0.98;
test_requires  'Test::Base';
test_requires  'Test::Fatal';
test_requires  'Test::Warn';

cc_assert_lib(
    incpath  => $INC_PATH,
    libpath  => $LIB_PATH,
    lib      => 'kytea',
    function => 'return 0;',

    # Devel::CheckLib does not find C++ compilers.
    #header   => [qw|kytea/kytea.h kytea/kytea-struct.h|],

    #debug    => 1,
);

check_version( fetch_version() );
train_kytea();

rewrite_model_path() if $INSTALL_DIR ne '/usr/local';

requires_cplusplus;
cc_src_paths 'xs';
#cc_warnings;
cc_libs "-L$LIB_PATH -lkytea";
cc_include_paths $INC_PATH;

use_ppport 3.20;

tests        't/*.t';
author_tests 'xt';

resources(
    license    => 'http://dev.perl.org/licenses',
    repository => 'https://github.com/pawa-/Text-KyTea',
    bugtracker => 'https://github.com/pawa-/Text-KyTea/issues',
);

WriteAll;


sub train_kytea
{
    unlink $TEST_MODEL;

    print "\n";
    print "Training kytea...\n";

    my $ret = system("$BIN_PATH/train-kytea -full $TEST_MODEL_FILE -model $TEST_MODEL");

    die "Is '$INSTALL_DIR' the correct path to the install directory of KyTea?" if $ret != 0;

    print "Training kytea done!\n";
    print "\n";
}

sub fetch_version
{
    my $version = qx($BIN_PATH/kytea --version 2>&1);
    chomp $version;

    $version =~ s/[^0-9]//g;
    $version =~ s/^0+//;

    return $version;
}

sub check_version
{
    my $version = shift;

    if ($version < $MIN_VERSION_INT)
    {
        $version = qx($BIN_PATH/kytea --version 2>&1) || 'version not found';
        chomp $version;

        die "This version of Text::KyTea supports KyTea version $MIN_VERSION and later."
          . " The version of your KyTea: $version";
    }
}

sub rewrite_model_path
{
    my $pm_path = './lib/Text/KyTea.pm';

    open(my $in, '<', $pm_path) or die $!;
    my @lines = <$in>;
    close($in);

    for my $line (@lines)
    {
        if ($line =~ m|model\s+=>\s+'/usr/local/share/kytea/model.bin'|)
        {
            $line = qq|        model => "$INSTALL_DIR/share/kytea/model.bin",\n|;
            last;
        }
    }

    open(my $out, '>', $pm_path) or die $!;
    print {$out} @lines;
    close($out);
}


lib/Text/KyTea.pm
package Text::KyTea;

use 5.008_001;
use strict;
use warnings;

use Carp ();

our $VERSION = '0.43';

require XSLoader;
XSLoader::load(__PACKAGE__, $VERSION);


sub _options
{
    return {
        # analysis options
        model   => '/usr/local/share/kytea/model.bin',
        nows    => 0,
        notags  => 0,
        notag   => [],
        nounk   => 0,
        unkbeam => 50,

        # I/O options
        tagmax  => 3,
        deftag  => 'UNK',
        unktag  => '',

        # advanced I/O options
        wordbound => ' ',
        tagbound  => '/',
        elembound => '&',
        unkbound  => ' ',
        skipbound => '?',
        nobound   => '-',
        hasbound  => '|',

        # Text::KyTea options
        prontag_num => 1,
    };
}

sub new
{
    my $class = shift;
    my %args  = (ref $_[0] eq 'HASH' ? %{$_[0]} : @_);

    my $options = $class->_options;

    for my $key (keys %args)
    {
        if ( ! exists $options->{$key} ) { Carp::croak "Unknown option: '$key'";  }
        else                             { $options->{$key} = $args{$key};        }
    }

    Carp::croak 'model file not found' unless -e $options->{model};

    return _init_text_kytea($class, $options);
}

1;

__END__

=encoding utf8

=head1 NAME

Text::KyTea - Perl wrapper for KyTea

=for test_synopsis
my ($text, %config);

=head1 SYNOPSIS

  use Text::KyTea;

  my $kytea   = Text::KyTea->new(%config);
  my $results = $kytea->parse($text);

  for my $result (@{$results})
  {
      print $result->{surface};

      for my $tags (@{$result->{tags}})
      {
          print "\t";

          for my $tag (@{$tags})
          {
              print " ", $tag->{feature}, "/", $tag->{score};
          }
      }

      print "\n";
  }


=head1 DESCRIPTION

KyTea is a general toolkit developed for analyzing text,
with a focus on Japanese, Chinese and other languages
requiring word or morpheme segmentation.

This module works under KyTea Ver.0.4.2 and later.
Under the old versions of KyTea, this module does not work!

If you've changed the default install directory of KyTea,
please install Text::KyTea in interactive mode
(e.g., cpanm --interactive or cpanm -v).

For more information about KyTea, please see the "SEE ALSO" section of this page.


=head1 METHODS

=head2 $kytea = Text::KyTea->new(%config)

Creates a new Text::KyTea instance.

  my $kytea = Text::KyTea->new(
      model       => 'model.bin', # default is "$INSTALL_DIRECTORY/share/kytea/model.bin"
      notag       => [1,2],       # default is []
      nounk       => 0,           # default is 0 (Estimates the pronunciation of unkown words.)
      unkbeam     => 50,          # default is 50
      tagmax      => 3,           # default is 3
      deftag      => 'UNK',       # default is 'UNK'
      unktag      => '',          # default is ''
      prontag_num => 1,           # default is 1 (Normally no need to change the default value.)
  );


=head2 $results_arrayref = $kytea->parse($text)

Parses the given text via KyTea, and returns the results.
The results are returned as an array reference.


=head2 $pron = $kytea->pron( $text [, $replacement ] )

Returns the estimated pronunciation of the given text.
Unknown pronunciations are replaced with $replacement.

If $replacement is not specified, unknown pronunciations
are replaced with the original characters.

This is the mere shortcut method.


=head2 $kytea->read_model($path)

Reads the given model file.
The model file should be read by new(model => $path) method.

Model files are available at L<http://www.phontron.com/kytea/model.html>.


=head1 AUTHOR

pawa E<lt>pawapawa@cpan.orgE<gt>

=head1 SEE ALSO

L<http://www.phontron.com/kytea/>

=head1 LICENSE

Copyright (C) pawa All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
